lld/ELF/Arch/EraVM.cpp

```diff
@@ -52,17 +52,37 @@ void EraVM::relocate(uint8_t *loc, const Relocation &rel, uint64_t val) const {
     write16be(loc, read16be(loc) + scaledValue);
   };
 
+  uint64_t scale = 0;
   switch (rel.type) {
   case R_ERAVM_16_SCALE_32:
-    add16scaled(val, /*scale=*/32);
+    scale = 32;
     break;
   case R_ERAVM_16_SCALE_8:
-    add16scaled(val, /*scale=*/8);
+    scale = 8;
     break;
-  default:
+  default: {
     error(getErrorLocation(loc) + "unrecognized relocation " +
           toString(rel.type));
+    return;
   }
+  }
+
+  // HACK: there may be cases where the relocation represents the code offset
+  // (measured in 8-byte units), for example:
+  //
+  //     add   @.OUTLINED_FUNCTION_RET0[0], r0, stack-[1]
+  //     jump  @OUTLINED_FUNCTION_0
+  //   .OUTLINED_FUNCTION_RET0:
+  //     jump.ge @.BB1_16
+  //     jump  @.BB1_23
+  //
+  // In such cases the actual relocation type should be R_ERAVM_16_SCALE_8.
+  // TODO: check if this can be done properly done on the LLVM MC layer?
+  auto *sec = cast<InputSection>(cast<Defined>(rel.sym)->section);
+  if (sec->name == ".text" && rel.type == R_ERAVM_16_SCALE_32)
```

Comment:
Instead of checking section name, maybe it is safer to check if section flags have `SHF_EXECINSTR`?
Also, regarding fix in the LLVM MC layer, maybe we can do something similar to [this](https://github.com/matter-labs/era-compiler-llvm/blob/9e9009229ea05aca97df5695933260cb2d78d010/llvm/lib/Target/BPF/MCTargetDesc/BPFELFObjectWriter.cpp#L54)?

Reply:
Hey @vladimirradosavljevic, thank you for the link! Will check it.

Reply:
Implemented the relocation type adjustment in the MC layer.